pkg/natdiscovery/request_handle.go

```diff
@@ -0,0 +1,156 @@
+		klog.Infof("received a NAT discovery packet from an cluster %q we don't know about yet", req.Receiver.GetClusterId())
+
+		response.Response = proto.ResponseType_UNKNOWN_SRC_CLUSTER
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if !nd.isKnownEndpoint(req.Sender.GetEndpointId()) {
+		// This can be ok for some time, because one cluster could discover an endpoint earlier on the broker than the
+		// the other endpoint has discovered him
+		klog.Infof("received a NAT discovery packet from an endpoint %q we don't know about yet", req.Receiver.GetEndpointId())
+
+		response.Response = proto.ResponseType_UNKNOWN_SRC_ENDPOINT
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if req.UsingSrc.GetIP() != "" && req.UsingSrc.GetIP() != addr.IP.String() {
+		klog.V(log.DEBUG).Infof("received NAT packet from endpoint %q, cluster %q, where NAT traversal has been detected, "+
+			"source IP changed",
+			req.Sender.GetEndpointId(), req.Sender.GetClusterId())
+		klog.V(log.DEBUG).Infof("original src IP was %q, received src IP is %q", req.UsingSrc.IP, addr.IP.String())
+
+		response.Response = proto.ResponseType_SRC_MODIFIED
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	if int(req.UsingSrc.Port) != addr.Port {
+		klog.V(log.DEBUG).Infof("received NAT packet from endpoint %q, cluster %q, where NAT traversal on the source has been detected, "+
+			"src port changed",
+			req.Sender.GetEndpointId(), req.Sender.GetClusterId())
+		klog.V(log.DEBUG).Infof("original src IP was %q, received src IP is %q", req.UsingSrc.IP, addr.IP.String())
+
+		response.Response = proto.ResponseType_SRC_MODIFIED
+
+		return nd.sendResponseToAddress(&response, addr)
+	}
+
+	response.Response = proto.ResponseType_OK
+
+	return nd.sendResponseToAddress(&response, addr)
+}
+
+func (nd *natDiscovery) isKnownCluster(clusterID string) bool {
+	nd.Lock()
+	defer nd.Unlock()
+
+	// NOTE: Iterating over map key to avoid copylock
+	for ep := range nd.remoteEndpoints {
+		if nd.remoteEndpoints[ep].Endpoint.Spec.ClusterID == clusterID {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (nd *natDiscovery) isKnownEndpoint(endpointID string) bool {
+	nd.Lock()
+	defer nd.Unlock()
+
+	_, ok := nd.remoteEndpoints[endpointID]
+
+	return ok
+}
+
+func (nd *natDiscovery) sendResponseToAddress(response *proto.SubmarinerNatDiscoveryResponse, addr *net.UDPAddr) error {
+	msgResponse := proto.SubmarinerNatDiscoveryMessage_Response{Response: response}
+	message := proto.SubmarinerNatDiscoveryMessage{Message: &msgResponse}
+	buf, err := proto2.Marshal(&message)
+	if err != nil {
+		return errors.Wrapf(err, "error marshaling response %#v", response)
+	}
+
+	if length, err := nd.serverUDPWrite(buf, addr); err != nil {
+		return errors.Wrapf(err, "error sending response packet %#v", response)
+	} else if length != len(buf) {
+		return errors.Errorf("the sent UDP packet was smaller than requested, sent=%d, expected=%d", length, len(buf))
```

Comment:
I'm yet to take a close look at the code. A quick question in the meantime - are we retrying when only partial data is sent?

Reply:
It will only be retried when we receive another request packet. The sender will retry.